libs/vault/src/services/copy-cipher-field.service.ts

```diff
@@ -0,0 +1,142 @@
+import { CipherView } from "@bitwarden/common/vault/models/view/cipher.view";
+import { ToastService } from "@bitwarden/components";
+import { PasswordRepromptService } from "@bitwarden/vault";
+
+/**
+ * The types of fields that can be copied from a cipher.
+ */
+export type CopyAction =
+  | "username"
+  | "password"
+  | "totp"
+  | "cardNumber"
+  | "securityCode"
+  | "email"
+  | "phone"
+  | "address"
+  | "secureNote"
+  | "hiddenField";
+
+type CopyActionInfo = {
+  /**
+   * The i18n key for the type of field being copied. Will be used to display a toast message.
+   */
+  typeI18nKey: string;
+
+  /**
+   * Whether the field is protected and requires password re-prompting before being copied.
+   */
+  protected: boolean;
+
+  /**
+   * Optional event to collect when the field is copied.
+   */
+  event?: EventType;
+};
+
+const CopyActions: Record<CopyAction, CopyActionInfo> = {
+  username: { typeI18nKey: "username", protected: false },
+  password: {
+    typeI18nKey: "password",
+    protected: true,
+    event: EventType.Cipher_ClientCopiedPassword,
+  },
+  totp: { typeI18nKey: "verificationCodeTotp", protected: true },
+  cardNumber: { typeI18nKey: "number", protected: true },
+  securityCode: {
+    typeI18nKey: "securityCode",
+    protected: true,
+    event: EventType.Cipher_ClientCopiedCardCode,
+  },
+  email: { typeI18nKey: "email", protected: false },
+  phone: { typeI18nKey: "phone", protected: false },
+  address: { typeI18nKey: "address", protected: false },
+  secureNote: { typeI18nKey: "note", protected: false },
+  hiddenField: {
+    typeI18nKey: "value",
+    protected: true,
+    event: EventType.Cipher_ClientCopiedHiddenField,
+  },
+};
+
+@Injectable({
+  providedIn: "root",
+})
+export class CopyCipherFieldService {
+  constructor(
+    private platformUtilsService: PlatformUtilsService,
+    private toastService: ToastService,
+    private eventCollectionService: EventCollectionService,
+    private passwordRepromptService: PasswordRepromptService,
+    private totpService: TotpService,
+    private i18nService: I18nService,
+    private billingAccountProfileStateService: BillingAccountProfileStateService,
+  ) {}
+
+  /**
+   * Copy a field value from a cipher to the clipboard.
+   * @param valueToCopy The value to copy.
+   * @param actionType The type of field being copied.
+   * @param cipher The cipher containing the field to copy.
+   * @param skipReprompt Whether to skip password re-prompting.
+   */
+  async copy(
+    valueToCopy: string,
+    actionType: CopyAction,
+    cipher: CipherView,
+    skipReprompt: boolean = false,
+  ) {
+    const action = CopyActions[actionType];
+    if (
+      !skipReprompt &&
+      cipher.reprompt !== CipherRepromptType.None &&
+      action.protected &&
+      !(await this.passwordRepromptService.showPasswordPrompt())
+    ) {
+      return;
+    }
+
+    if (valueToCopy == null || !cipher.viewPassword) {
```

Comment:
I think we can combine this check here 

```suggestion
    if (valueToCopy == null || !cipher.viewPassword || (actionType === "totp" && !(await this.totpAllowed(cipher)))) {
```

Reply:
We could combine it, but we'd still need a separate check for `actionType === "totp"` in order to use the `totpService.getCode()` method below. I personally think its a little more clear to keep the `totp` "branch" separate  and have fewer conditions in a single `if` even if it repeats the empty `return` below.